src/besseli.jl

```diff
@@ -1,46 +1,105 @@
+# Modified Bessel functions of the first kind of order zero and one
+# besseli0, besseli1
+# Scaled modified Bessel functions of the first kind of order zero and one
+# besseli0x, besselix
+
 #=
-Cephes Math Library Release 2.8:  June, 2000
-Copyright 1984, 1987, 2000 by Stephen L. Moshier
-https://github.com/jeremybarnes/cephes/blob/master/bessel/i0.c
-https://github.com/jeremybarnes/cephes/blob/master/bessel/i1.c
+Approximate forms used here are given in
+"Rational Approximations for the Modified Bessel Function of the First Kind - I1(x) for Computations with Double Precision"
+by Pavel Holoborodko, 
+http://www.advanpix.com/2015/11/12/rational-approximations-for-the-modified-bessel-function-of-the-first-kind-i1-for-computations-with-double-precision/
+Actual coefficients used are from the boost math library.
+https://github.com/boostorg/math/tree/develop/include/boost/math/special_functions/detail
 =#
+
 function besseli0(x::T) where T <: Union{Float32, Float64}
-    if x <= 8
-        y = muladd(x, T(.5), T(-2))
-        return exp(x) * chbevl(y, A_i0(T))
+    T == Float32 ? branch = 50 : branch = 500
+    x = abs(x)
+    if x < 7.75
+        a = x * x / 4
+        return muladd(a, evalpoly(a, P1_i0(T)), 1)
+    elseif x < branch
+        return exp(x) * evalpoly(inv(x), P2_i0(T)) / sqrt(x)
```

Comment:
I don't think this branch is worthwhile the branch for large `x` handles medium `x` well (and cheaply). Also, recalculating these coefficients so you are using `exp2` rather than `exp` would save you a small amount of time and precision.

Reply:
The issue with `besseli` and `besselk` is their diverging properties. This small branch is to prevent overflow in exp(x) before there is overflow in `besseli`. 
So for example.
```julia
### using SpecialFunctions
julia> besseli(0, 710.0)
ERROR: AmosException with id 2: overflow.
Stacktrace:
 [1] _besseli(nu::Float64, z::ComplexF64, kode::Int32)
......

##### using this code
julia> besseli0(710.0)
3.3453345586196564e306
```
So we can see that the `besseli` implementation in SpecialFunctions.jl does not handle this zone properly and overflows before `besseli` actually does. So it is a rather small branch but it is able to provide more accurate results when you reach closer to floatmax.

Do you think this is worth it?

Reply:
what I'm saying is to use the branch that you use for large numbers for all numbers, since the `exp(x/2)` will work for the whole range.

Reply:
Ahh I see! Ok so if we change it by removing the `P3_i0` branch
```julia
function besseli0(x::T) where T <: Union{Float32, Float64}
    x = abs(x)
    if x < 7.75
        a = x * x / 4
        return muladd(a, evalpoly(a, P1_i0(T)), 1)
    else
        a = exp(x / 2)
        s = a * evalpoly(inv(x), P2_i0(T)) / sqrt(x)
        return a * s
    end
end
```
This will pass all tests.
The trade off becomes `P2_i0` is length 22 whereas `P3_i0` is length 2 so large arguments will take a little more time to compute. 
Ok with that? I can do some benchmarks

Reply:
Some benchmarks
```julia
# removing the third branch
julia> @benchmark besseli0(x) setup=(x=rand(500:0.0001:700))
BenchmarkTools.Trial: 10000 samples with 996 evaluations.
 Range (min … max):  22.925 ns … 63.379 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     23.050 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   23.092 ns ±  0.629 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

               █     █      ▁                                  
  ▂▁▁▁▁▁▆▁▁▁▁▁▁█▁▁▁▁▁█▁▁▁▁▁▁█▁▁▁▁▁█▁▁▁▁▁▁▅▁▁▁▁▁▃▁▁▁▁▁▁▂▁▁▁▁▁▂ ▂
  22.9 ns         Histogram: frequency by time        23.3 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

# keeping the third branch
julia> @benchmark besseli0(x) setup=(x=rand(500:0.0001:700))
BenchmarkTools.Trial: 10000 samples with 999 evaluations.
 Range (min … max):  12.053 ns … 25.985 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     12.137 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   12.186 ns ±  0.474 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

           █    ▁                                              
  ▂▁▁▁▄▃▁▁▁█▁▁▁▁█▁▁▁▁▄▁▁▁▁▃▁▁▁▁▂▁▁▁▁▂▁▁▁▁▂▁▁▁▁▁▁▁▁▁▂▁▁▁▁▂▁▁▁▂ ▂
  12.1 ns         Histogram: frequency by time        12.6 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

I also tested for smaller arguments and they are roughly similar.
```julia
julia> @benchmark besseli02(x) setup=(x=rand(0.00001:0.0001:100))
BenchmarkTools.Trial: 10000 samples with 999 evaluations.
 Range (min … max):  10.594 ns … 48.674 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     22.690 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   21.880 ns ±  3.261 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                                                     █▁        
  ▄▂▁▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁██▂▂▂▂▂▂ ▂
  10.6 ns         Histogram: frequency by time        24.4 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark besseli0(x) setup=(x=rand(0.00001:0.0001:100))
BenchmarkTools.Trial: 10000 samples with 999 evaluations.
 Range (min … max):   9.635 ns … 40.666 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     22.731 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   21.822 ns ±  3.526 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▄                                                  ▃█  ▂    ▁
  █▇▄▄▁▃▁▁▁▁▁▁▁▁▁▃▁▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁███▅█▇▇▇ █
  9.63 ns      Histogram: log(frequency) by time      24.4 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

Reply:
I think that's probably a good compromise (but I'm not 100% sure and it probably depends on how frequently you call  the function with big arguments.

Reply:
To see the cost of the branch, you're going to want to test a mix of above and below (unsorted) so the branch predictor can't save you

Reply:
but your current version might just be the way to go.

Reply:
One thing I don't have experience in is optimizing this for SIMD better. It would be nice to have these play well with LoopVectorization.. Unfortunately, I've frequently had to compute bessel functions for large arguments. They are often used in integral transforms requiring a large number of evaluations to reach good convergence (but this could just be a very special case where `besseli(x)` is needed in an ordered array from 0 -> inf).

I'll do some more benchmarks and leave this open for a bit! If you have any resources for providing better SIMD-able functions let me know. It could be its well worth dropping this branch if optimizations can be made elsewhere.

Reply:
tldr is that for a simd implementation, you end up calculating both sides of the branch, so it will definitely be bad.

Reply:
I'll propose removing this branch for `besseli0` and `besseli1` and keeping it for the scaled versions. In general, if you are anticipating large arguments you should be designing your code to use the scaled versions. So if you are using `besseli0x` you are much more likely to call it for very large arguments. Maybe next I can work on the docs 😅